Autopublication des blagues sur GitHub

achille-bsc opened the pull request:
Auto publication des blagues sur GitHub

Closes: #398

swiizyy commented on src/bot/index.ts:
```diff
@@ -148,6 +164,27 @@ export default class Bot extends Client {
     await this.login(process.env.BOT_TOKEN);
   }
+
+  async runGitPush(): Promise<void> {
+    console.log('Push lancée avec succès !');
+
+    try {
+      gitCommitPush({
+        owner: 'blagues-api',
+        repo: 'blagues-api',
+        // commit files
+        file: { path: 'blagues.json', content: fs.readFileSync('./blagues.json', 'utf-8') },
+        baseBranch: 'dev',
+        mergeBranch: 'test'
+      });
+    } catch (error) {
+      console.log(error);
+      const channel = this.channels.cache.get(process.env.LOGS_CHANNEL!) as TextChannel;
+      await channel.send({
+        ...interactionProblem(`L'erreur suivante est survenur: \`\`\`${error}\`\`\``)
+      });
+    }
+  }
```
Personnellement je ne l'aurais pas mis là mais dans une classe comme pour les autres modules 👀
Tu en pense quoi ?

DraftProducts replied:
Oui 👍

swiizyy commented on src/bot/modules/push.ts:
```diff
@@ -0,0 +1,110 @@
+      commit_message: 'Merge des dernières blagues'
+    });
+  } catch (error) {
+    console.error(error);
+    // TODO: Créer la pull request uniquement si le code d'erreur match bien
+    await octokit.pulls.create({
+      owner: options.owner,
+      repo: options.repo,
+      title: "[Blagues] Conflit lors de l'ajout des dernières blagues",
+      head: options.mergeBranch,
+      base: options.baseBranch,
+      body: `\`\`\`${error}\`\`\``,
+      maintainer_can_modify: true
+    });
+  }
+}
+
+export interface GitCommitPushOptions {
+  owner: string;
+  repo: string;
+  file: {
+    path: string;
+    content: string;
+  };
+  baseBranch: string;
+  mergeBranch: string;
+}
+
+export async function gitCommitPush(options: GitCommitPushOptions) {
+  const octokit = new Octokit({
+    auth: process.env.GITHUB_API_TOKEN
+  });
+
+  const referenceSha = await getReferenceCommit(octokit, options);
+  const treeSha = await createTree(octokit, options, referenceSha);
+  const commitSha = await createCommit(octokit, options, { referenceSha, treeSha });
+
+  const branchSha = await createReference(octokit, options, commitSha);
+  await mergeReference(octokit, options, branchSha);
+}
```
J'aurais plutôt mis sa dans le dossier lib 👀

DraftProducts replied:
Je suis d'accord 👍

DraftProducts commented on .env.example:
```diff
@@ -1,6 +1,7 @@
 BASE_URL=http://localhost:3000
 
 BOT_TOKEN=
+GITHUB_API_TOKEN=
```
Entre `JOKES_AUTO_PUBLICATION` et `SENTRY_DSN` avec un retour a la ligne de chaque coté

DraftProducts commented on .vscode/settings.json:
```diff
@@ -15,5 +15,10 @@
   "deepscan.enable": true,
-  "conventionalCommits.scopes": ["bot", "api", "nuxt", "ts"]
-}
+  "conventionalCommits.scopes": [
+    "bot",
+    "api",
+    "nuxt",
+    "ts"
+  ]
+}
```
Je ne comprends pas pourquoi tu rechanges ça à chaque fois, c'est formaté à partir des règles du projet, essayes de régler le conflit dans ta config personnelle ou ne commit pas ce fichier s'il te plait.

DraftProducts commented on src/bot/commands/approve.ts:
```diff
@@ -43,7 +43,10 @@ import { compareTwoStrings } from 'string-similarity';
   constructor() {
     super({
-      name: 'Approuver',
+      name: 'Approve',
+      nameLocalizations: {
+        fr: 'Approuver'
+      },
```
Pourquoi a-t-on ces changements de nouveau ? Ces changements n'ont aucun lien avec la PR et ont été merge d'une autre PR, il y a déjà un moment.

DraftProducts commented on src/bot/index.ts:
```diff
@@ -108,7 +121,10 @@ export default class Bot extends Client {
           if (fetchedMessage.deletable) await fetchedMessage.delete();
-        } catch {}
+        } catch {
+          // Je remet le continue par ce que sinon ya une erreur entre prettier et eslint
```
```suggestion
```

DraftProducts commented on src/bot/index.ts:
```diff
@@ -59,6 +64,14 @@ export default class Bot extends Client {
     await this.dispatcher.register();
     await this.stickys.reload();
 
+    // await this.runGitPush();
+    if (process.env.JOKES_AUTO_PUBLICATION === 'true') {
+      // 0 0 * * 1
+      // At 00:00 on Monday
+      schedule.scheduleJob('0 0 * * 1', this.runGitPush); // On execute runGitPush() tout  les dimanche à 21H => https://www.npmjs.com/package/node-schedule
+    } else {
+      console.log('Auto publication des blagues sur gitHub désactivé !');
+    }
```
```suggestion
    if (process.env.JOKES_AUTO_PUBLICATION === 'true') {
      // At 00:00 on Monday
      schedule.scheduleJob('0 0 * * 1', this.runGitPush);
    } else {
      console.log('Service d\'auto-publication des blagues sur GitHub désactivé');
    }
```

DraftProducts reviewed (changes requested):
Avant de commit tes fichiers, pense à bien vérifier que tu es d'accord avec tous les changements que tu vas apporter.
- Là, je vois que tu as changé tous les liens des dépendances dans le fichier `yarn.lock`.
- Que tu aies commit des changements qui n'étaient pas du ressort de cette PR (et qui sont d'ailleurs déjà dans la codebase) c'est du au fait qu'encore une fois tu n'as pas créé la branche à partir de dev mais à a partir d'une PR qui a été squash.
- Le reformatage des fichiers de config ect...
- Tous les commentaires de tests sont à enlever avant tes commits
- Si tu souhaites justifier un changement comme le continue par exemple, fait le dans la PR, mais pas dans le code.

Je suis d'accord avec ce qu'a proposé @Swiizyy 👍 
Certaines parties du code ne sont pas à la bonne place.

DraftProducts committed:
feat: ✨ Add GitHub methods to commit/merge/push

Co-authored-by: AchilleBosc <discordtopinfos@gmail.com>

DraftProducts committed:
feat: ✨ Add cron automatization to auto push changes

Co-authored-by: AchilleBosc <discordtopinfos@gmail.com>

DraftProducts commented:
Je viens de réécrire l'historique pour que tu repartes d'une bonne base :)
Pense juste à pull et après, tu peux commencer à corriger, à apporter les modifications de la review

ErwanGit commented on src/bot/index.ts:
```diff
@@ -148,6 +164,27 @@ export default class Bot extends Client {
+      console.log(error);
+      const channel = this.channels.cache.get(process.env.LOGS_CHANNEL!) as TextChannel;
+      await channel.send({
+        ...interactionProblem(`L'erreur suivante est survenur: \`\`\`${error}\`\`\``)
```
```suggestion
        ...interactionProblem(`Une erreur suivante est survenue : \`\`\`${error}\`\`\``)
```

achille-bsc replied:
Ah bon ? ça me surprend un peu... ça ne me parait pas vraiment correct comme phrase, si ? Ça fait un peu bizarre à lire, je trouve, non ?

achille-bsc committed:
Update src/bot/index.ts

Co-authored-by: Erwan <24718500+ErwanGit@users.noreply.github.com>

Commit:
build(bot): 🔥 Move push file to the lib folder

Commit:
build(bot): 🔥 Remove and move comments and functions

Remove useless code comments and move runGitPush function to an other file in module folder

DraftProducts reviewed (changes requested):
Regarde le module dans `reminders.ts`, il faudrait que ça ait cette forme, ça permettrait d'enlever le `node-schedule` du fichier index du bot.

achille-bsc commented:
> Regarde le module dans `reminders.ts`, il faudrait que ça ait cette forme, ça permettrait d'enlever le `node-schedule` du fichier index du bot.

Je ne suis paq sur de savoir ce que tu veux dire par là 😅

Swiizyy commented:
> > Regarde le module dans `reminders.ts`, il faudrait que ça ait cette forme, ça permettrait d'enlever le `node-schedule` du fichier index du bot.
> 
> Je ne suis paq sur de savoir ce que tu veux dire par là 😅

Comme ceci :)


```ts
// AutoPublish.ts
import { Client } from 'discord.js';
import schedule from 'node-schedule';

export default class Reminders {
  public client: Client;

  constructor(client: Client) {
    this.client = client;

    if (process.env.JOKES_AUTO_PUBLICATION === 'false') return;

    // Every 10 minutes
    schedule.scheduleJob('0 0 * * 1', () => this.run());
  }

  async run(date: Date): Promise<void> {
    // Ajouter le code de ta fonction runGitPush
  }
}
```
```ts
// index.ts
import AutoPublish from './modules/autoPublish';

export default class Bot extends Client {
  public autoPublish: AutoPublish
  constructor() {
    this.autoPublish = new AutoPublish(this)
  }
}
```

swiizyy commented on src/bot/modules/autoPublication.ts:
```diff
@@ -0,0 +1,26 @@
+    });
+  } catch (error) {
+    console.log(error);
+    const channel = bot.channels.cache.get(process.env.LOGS_CHANNEL!) as TextChannel;
```
Est tu sur de vouloir recevoir le message d'erreur avec tout les autres logs ?
J'aurais plutot vu un DM a DraftMan pour ce cas de figure :)

achille-bsc replied:
Non, par ce que en cas d'erreur, il pourrait être spam en mp, ce qui ne serait pas désirable... alors que les logs sont plus aptes à recevoir des erreurs

swiizyy replied:
Non pas du tout c'est dans un try catch donc il serra avertie une seul fois nan ? 🧐

achille-bsc replied:
Bah nan, un try catch n'arrête pas le code, ça permet justement de gérer une erreur sans faire crash le programme, donc il y a toujours la possibilité que ça refasse la même erreur plus tard je crois

DraftProducts replied:
Je serais plus pour qu'on envoie ce genre de message dans la console 🤔 
un simple `console.error` suffirait.

swiizyy commented on src/bot/modules/autoPublication.ts:
```diff
@@ -0,0 +1,26 @@
+      // commit files
+      file: { path: 'blagues.json', content: fs.readFileSync('./blagues.json', 'utf-8') },
+      baseBranch: 'dev',
+      mergeBranch: 'test'
```
```suggestion
      mergeBranch: 'chore/autopublish-jokes'
```

swiizyy commented on src/bot/modules/autoPublication.ts:
```diff
@@ -0,0 +1,26 @@
+import Bot from 'bot';
+
+export async function runGitPush(bot: Bot): Promise<void> {
+  console.log('Push lancée avec succès !');
```
```suggestion
```

achille-bsc committed:
Update src/bot/modules/autoPublication.ts

Co-authored-by: Williams Bussat <williams@wbussat.fr>

achille-bsc committed:
Update src/bot/modules/autoPublication.ts

Co-authored-by: Williams Bussat <williams@wbussat.fr>

ErwanGit commented on src/bot/index.ts:
```diff
@@ -59,6 +61,14 @@ export default class Bot extends Client {
+        runGitPush(this);
+      }); // On execute runGitPush() tout  les dimanche à 21H => https://www.npmjs.com/package/node-schedule
+    } else {
+      console.log('Auto publication des blagues sur gitHub désactivé !');
```
```suggestion
      console.log('Auto-publication des blagues sur GitHub désactivée');
```

ErwanGit commented on src/bot/modules/autoPublication.ts:
```diff
@@ -0,0 +1,25 @@
+    console.log(error);
+    const channel = bot.channels.cache.get(process.env.LOGS_CHANNEL!) as TextChannel;
+    await channel.send({
+      ...interactionProblem(`L'erreur suivante est survenur: \`\`\`${error}\`\`\``)
```
```suggestion
      ...interactionProblem(`L'erreur suivante est survenue : \`\`\`${error}\`\`\``)
```

achille-bsc committed:
Update src/bot/index.ts

Co-authored-by: Erwan <24718500+ErwanGit@users.noreply.github.com>

achille-bsc committed:
Update src/bot/modules/autoPublication.ts

Co-authored-by: Erwan <24718500+ErwanGit@users.noreply.github.com>

DraftProducts commented:
> Comme ceci :)
> 
> ```ts
> // AutoPublish.ts
> import { Client } from 'discord.js';
> import schedule from 'node-schedule';
> 
> export default class Reminders {
>   public client: Client;
> 
>   constructor(client: Client) {
>     this.client = client;
> 
>     if (process.env.JOKES_AUTO_PUBLICATION === 'false') return console.log('Service d\'auto-publication des blagues sur GitHub désactivé');
> 
>     // Every 10 minutes
>     schedule.scheduleJob('0 0 * * 1', () => this.run());
>   }
> 
>   async run(date: Date): Promise<void> {
>     // Ajouter le code de ta fonction runGitPush
>   }
> }
> ```
> 
> ```ts
> // index.ts
> import AutoPublish from './modules/autoPublish';
> 
> export default class Bot extends Client {
>   public autoPublish: AutoPublish
>   constructor() {
>     this.autoPublish = new AutoPublish(this)
>   }
> }
> ```

👍

swiizyy committed:
refactor: :recycle: Refactor gitPushCommit to modules class

DraftProducts commented on src/bot/index.ts:
```diff
@@ -45,6 +44,13 @@ export default class Bot extends Client {
     this.stickys = new Stickys(this);
     this.reminders = new Reminders(this);
     this.votes = new Votes(this);
+    this.autoPublish = new AutoPublish(this, {
+      owner: process.env.GITHUB_OWNER ?? 'blagues-api',
+      repo: process.env.GITHUB_REPO ?? 'blagues-api',
+      file: { path: 'blagues.json', content: readFileSync('./blagues.json', 'utf-8') },
+      baseBranch: process.env.GITHUB_BASE_BRANCH ?? 'dev',
+      mergeBranch: process.env.GITHUB_MERGE_BRANCH ?? 'chore/autopublish-jokes'
+    });
```
Bonne idée mais je serais plus pour l'avoir dans le module directement pour éviter de surcharger ce fichier.

swiizyy replied:
J'ai vachement hésité sur se coup la mais oui tu a totalement raison :)

DraftProducts reviewed (approved):
Très beau boulot et très bonnes initiatives mais même si @achille-bsc mettait du temps j'aurais préféré que tu lui laisses s'en occuper étant donné que c'était sa PR (sauf si tu as eu son accord au préalable).

swiizyy committed:
fix: 🐛 Fix from review

achille-bsc commented:
> Très beau boulot et très bonnes initiatives mais même si @achille-bsc mettait du temps j'aurais préféré que tu lui laisses s'en occuper étant donné que c'était sa PR (sauf si tu as eu son accord au préalable).

Il a effectivement eu mon accord au préalable ;) sinon je m'en serait occupé cet aprem, mais au moins c'est fait !
D'ailleurs je tien à m'excuser pour le temps que je vous ai tous fais perdre en n'ayant pas relu mon code ! J'ai retenu la leçon et ferait mieux la prochaine fois !

DraftProducts commented:
Pour moi tu peux merge 🔥 @Swiizyy

Swiizyy commented:
J'ai juste une petite suggestion avant de merge @DraftProducts :)

Une petite commande pour que tu puisse merge les blagues au cas ou avant le dimanche serait intéressant ou pas  ?

swiizyy committed:
feat: ♻️ Improve response with mergeBranchStatus

DraftProducts commented on src/bot/modules/autoPublication.ts:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  #responseWithStatus: Record<Status, () => void> = {
+    '201': () => console.log('[Auto-Publish] La publication des blagues sur Github effectué avec succés !'),
+    '204': () => console.log('[Auto-Publish] La publication des blagues sur Github déja effectué !'),
+    '403': () =>
+      console.log(
+        `[Auto-Publish] Je n'est malheuresement pas la permission d'accéder au repertoire suivant: ${this.options.owner}/${this.options.repo}`
+      ),
+    '404': () => console.error('[Auto-Publish] Branche ou tête introuvable !'),
+    '409': () => {
+      this.octokit.pulls.create({
+        owner: this.options.owner,
+        repo: this.options.repo,
+        title: "[Blagues] Conflit lors de l'ajout des dernières blagues",
+        head: this.options.mergeBranch,
+        base: this.options.baseBranch,
+        body: stripIndents`
+          ⚠️ Conflit détectée lors de l'ajout des dernières blagues !
+
+          Veuillez résoudre ces conflits afin de fusionner la branche manuellement :)
+        `,
+        maintainer_can_modify: true
+      });
+      console.error(
+        `[Auto-Publish] Conflit détécté lors de la publication des blagues sur Github !\n Une Pull-Request a été crée.`
+      );
+    },
+    '422': () =>
+      console.error('[Auto-Publish] La publication des blagues sur Github ou le point de terminaison a été spammé. !')
```
```suggestion
    '201': () => console.log('[Auto-Publish] Publication des blagues sur Github effectuée avec succès !'),
    '204': () => console.log('[Auto-Publish] Publication des blagues sur Github déjà effectuée !'),
    '403': () =>
      console.log(
        `[Auto-Publish] Permission d'accès manquante au repo "${this.options.owner}/${this.options.repo}"`
      ),
    '404': () => console.error(`[Auto-Publish] Branche "${this.options.baseBranch}" introuvable !`),
    '409': () => {
      this.octokit.pulls.create({
        owner: this.options.owner,
        repo: this.options.repo,
        title: "[Blagues] Conflit lors de l'ajout des dernières blagues",
        head: this.options.mergeBranch,
        base: this.options.baseBranch,
        body: stripIndents`
          ⚠️ Conflit détectée lors de l'ajout des dernières blagues !

          Veuillez résoudre ces conflits afin de fusionner la branche manuellement :)
        `,
        maintainer_can_modify: true
      });
      console.error(
        `[Auto-Publish] Conflit détecté lors de la publication des blagues sur Github !\n Une Pull-Request a été créée.`
      );
    },
    '422': () =>
      console.error('[Auto-Publish] Un ratelimit est survenu, la publication des blagues sur Github a été reportée à dimanche prochain !')
```

DraftProducts commented on src/bot/modules/autoPublication.ts:
```diff
@@ -0,0 +1,155 @@
+    return this.octokit.rest.repos.merge({
+      owner: this.options.owner,
+      repo: this.options.repo,
+      base: `${this.options.baseBranch}`,
```
```suggestion
      base: this.options.baseBranch,
```

DraftProducts reviewed (changes requested):
Très bonne idée et super boulot gg ! 🔥

swiizyy committed:
Update src/bot/modules/autoPublication.ts

Co-authored-by: DraftMan <contact@draftman.fr>

swiizyy committed:
Update src/bot/modules/autoPublication.ts

Co-authored-by: DraftMan <contact@draftman.fr>

Swiizyy commented:
> Très bonne idée et super boulot gg ! 🔥 

Merci 😁

DraftProducts commented:
Eeeeh

DraftProducts commented:
Kesissépassé ? @Swiizyy

Swiizyy commented:
> Kesissépassé ? @Swiizyy

Oupssi j'etait sur tel mdr :)

swiizyy committed:
Merge branch 'dev' into feat/joke-aut-publication

swiizyy committed:
fix: 🚨 Fix lint error

swiizyy committed:
fix: 🐛 Fix import stats modules

achille-bsc commented on src/bot/commands/userStats.ts:
```diff
@@ -1,6 +1,6 @@
 import Command from '../lib/command';
 import { ApplicationCommandType, UserContextMenuCommandInteraction } from 'discord.js';
-import Stats from '../modules/stats';
+import { Stats } from '../modules/stats';
```
```suggestion
import { Stats } from '../modules';
```
ce serait bien de faire cette petite modif dans tout les fichiers qui demande ce fichier
